src/codec.rs

```diff
@@ -664,19 +696,51 @@ impl Decode for Compact<u64> {
 		let prefix = input.read_byte()?;
 		Ok(Compact(match prefix % 4 {
 			0 => prefix as u64 >> 2,
-			1 => u16::decode(&mut PrefixInput{prefix: Some(prefix), input})? as u64 >> 2,
-			2 => u32::decode(&mut PrefixInput{prefix: Some(prefix), input})? as u64 >> 2,
+			1 => {
+				let x = u16::decode(&mut PrefixInput{prefix: Some(prefix), input})? >> 2;
+				if x > 0b00111111 && x <= 0b00111111_11111111 {
+					x as u64
+				} else {
+					return Err(U64_OUT_OF_RANGE.into());
+				}
+			},
+			2 => {
+				let x = u32::decode(&mut PrefixInput{prefix: Some(prefix), input})? >> 2;
+				if x > 0b00111111_11111111 && x <= u32::max_value() >> 2 {
+					x as u64
+				} else {
+					return Err(U64_OUT_OF_RANGE.into());
+				}
+			},
 			3|_ => match (prefix >> 2) + 4 {
-				4 => u32::decode(input)? as u64,
-				8 => u64::decode(input)?,
+				4 => {
+					let x = u32::decode(input)?;
+					if x > u32::max_value() >> 2 {
+						x as u64
+					} else {
+						return Err(U64_OUT_OF_RANGE.into());
+					}
+				},
+				8 => {
+					let x = u64::decode(input)?;
+					if x > u64::max_value() >> 8 {
```

Comment:
I think this is incorrect, if we require 8 bytes, we have a value that requires a full `u64`.

Reply:
So the encoded value actually requires 9 bytes.

Reply:
I am not sure I understand; this just means that on 8 byte an encoding happen iff it could not happen on 7.
So it happens if value is more than what could be encoded on 7 (as in the test) and less than what can be on 8 (we decode from 8 byte).
(when we match 8 here the actual encoding is 9 byte length but the read length from first byte is 8)

Reply:
I did ran 
```rust
	#[test]
	fn compact_u64_test() {
		for a in [
			u64::max_value(), 
			u64::max_value() - 1,
			u64::max_value() << 8,
			(u64::max_value() << 8) - 1,
			u64::max_value() << 16,
			(u64::max_value() << 16) - 1,
		].into_iter() {
			let e = Compact::<u64>::encode(&Compact(*a));
			let d = Compact::<u64>::decode(&mut &e[..]).unwrap().0;
			assert_eq!(*a, d);
		}
	}

	#[test]
	fn compact_u128_test() {
		for a in [
			u64::max_value() as u128, 
			(u64::max_value() - 10) as u128, 
			u128::max_value(),
			u128::max_value() - 10,
		].into_iter() {
			let e = Compact::<u128>::encode(&Compact(*a));
			let d = Compact::<u128>::decode(&mut &e[..]).unwrap().0;
			assert_eq!(*a, d);
		}
	}
```
without a pb (but as I say I may not have understood what you meant)